include/seqan3/alphabet/structure/structure_composition.hpp

```diff
@@ -0,0 +1,186 @@
+ * \tparam sequence_alphabet_t Type of the first letter; must satisfy seqan3::nucleotide_concept.
+ * \tparam structure_alphabet_t Types of further letters (up to 4); must satisfy seqan3::structure_concept.
+ *
+ * This composition pairs a nucleotide alphabet with a structure alphabet. The rank values
+ * correpsond to numeric values in the size of the composition, while the character values
+ * are taken from the sequence alphabet and the structure annotation is taken from the structure
+ * alphabet.
+ *
+ * As with all `seqan3::cartesian_composition` s you may access the individual alphabet letters in
+ * regular c++ tuple notation, i.e. `get<0>(t)` and objects can be brace-initialized
+ * with the individual members.
+ *
+ * ~~~~~~~~~~~~~~~{.cpp}
+ *
+ * structure_composition<rna4, dot_bracket3> l{rna4::G, dot_bracket3::PAIR_OPEN};
+ * std::cout << int(to_rank(l)) << ' '
+ *           << int(to_rank(get<0>(l))) << ' '
+ *           << int(to_rank(get<1>(l))) << '\n';
+ * // 6 2 1
+ *
+ * std::cout << to_char(l) << ' '
+ *           << to_char(get<0>(l)) << ' '
+ *           << to_char(get<1>(l)) << '\n';
+ * // G G (
+ *
+ * // modify via structured bindings and references:
+ * auto & [ seq_l, structure_l ] = l;
+ * seq_l = rna4::U;
+ * std::cout << to_char(l) << '\n';
+ * // U
+ *
+ * ~~~~~~~~~~~~~~~
+ *
+ * This seqan3::cartesian_composition itself fulfills both seqan3::alphabet_concept and seqan3::structure_concept .
+ */
+
+template <typename sequence_alphabet_t, typename structure_alphabet_t>
+      requires nucleotide_concept<sequence_alphabet_t> &&
+               structure_concept<structure_alphabet_t>
+struct structure_composition :
+    public cartesian_composition<structure_composition<sequence_alphabet_t, structure_alphabet_t>,
+                                 sequence_alphabet_t, structure_alphabet_t>
+{
+    //!\brief First template parameter as member type.
+    using sequence_alphabet_type = sequence_alphabet_t;
+    //!\brief Second template parameter as member type.
+    using structure_alphabet_type = structure_alphabet_t;
+
+    //!\brief Equals the char_type of sequence_alphabet_type.
+    using char_type = underlying_char_t<sequence_alphabet_type>;
+    //!\brief Equals the char_type of the structure_alphabet_type.
+    using str_char_type = underlying_char_t<structure_alphabet_type>;
+
+    /*!\name Write functions
+     * \{
+     */
+    //!\brief Directly assign the sequence character.
+    constexpr structure_composition & operator=(sequence_alphabet_type const l) noexcept
+    {
+        get<0>(*this) = l;
+        return *this;
+    }
+
+    //!\brief Directly assign the structure character.
+    constexpr structure_composition & operator=(structure_alphabet_type const l) noexcept
+    {
+        get<1>(*this) = l;
+        return *this;
+    }
+
+    //!\brief Assign from a nucleotide character. This modifies the internal sequence letter.
+    constexpr structure_composition & assign_char(char_type const c)
```

Comment:
No, you are wrong this time. cartesian_composition is built using CRTP. It is explicitly designed to return the derived class on assignment.